### DISCUSSION TITLE
vjunos-switch on BMS

### DISCUSSION THREAD

**Discussion by sjohnson1980** [ipspace/netlab#1253] [Features]
Hi,

I looked at the supported platforms and noticed that vjunos-switch wasn't listed. I know that there are possible issues with nested virtualization, but if I was to run netlab on a BMS, using libvirt, just as it can be done with EVE or Containerlab, would it be supported? Or is this tied with Vagrant proper support or to prevent nested virtualization help requests?

Thanks

**Comment by ipspace** [comment]
Can't remember what it was that made us decide to go with vPTX. vjunos-switch might have required two VMs when they first made it available, or might have some weird nested virtualization requirements.

It looks like it's a single VM now, and The Internet claims Vagrant should be able to do nested virtualization, so I'm guessing adding support for it seems like a no-brainer. All we need now is a volunteer ;)

If only someone could push their vendor rep hard enough to make it happen ;))

/cc @ssasso @nmodena

**Reply by sjohnson1980** [reply]
Indeed, vjunos-switch is a VM that contains 2 KVM VMs, one for the RE and one for the PFE, that are nested into a single image. Hence the requirement of a BMS to avoid dual nested virtualization issues. It's not recommended to deploy this image on a VM from any hypervisor.

**Reply by ipspace** [reply]
Ah, now it all came back 🤢

We could do it as long as the integration effort is reasonably small (= changing interface names and such).

**Reply by ssasso** [reply]
indeed, I spoke to my contacts using it and they are all using containerlab on a bare metal server.

adding the "code" for this new kind on netlab should be straightforward: @sjohnson1980 if I create a dedicated branch for this, would you be able to test it with containerlab in your bare metal environment using containerlab?

Unfortunately I don't have a BM that I can use right now, and I started a clab vjunos-switch instance on my VM 2 hours ago and not yet able to ssh or serial login ;)

**Reply by ssasso** [reply]
side note: now the VM is up... but... https://github.com/hellt/vrnetlab/issues/231

**Reply by sjohnson1980** [reply]
Thanks for the response ssasso. Yes, I'd be able to test it but not until at least next week because of current engagements, even if I'd be interested.

**Reply by ipspace** [reply]
I have a bare-metal server, and got the container running in somewhat-reasonable time. Even SSH works.

Give me a day or two, and I'll probably have some baseline implementation.

**Reply by ssasso** [reply]
@ipspace I'm committing/PR the baseline in a few minutes ;)

**Reply by ipspace** [reply]
OK, waiting ;))

**Comment by ssasso** [comment]
I know different people running vjunos-switch in containerlab, so it should be an easy task integrating it in netlab.

There might be some customization needed for the way vlans are handled, most probably, but all the other modules should work as is.

**Reply by ipspace** [reply]
> I know different people running vjunos-switch in containerlab, so it should be an easy task integrating it in netlab.

Does this sound like volunteering? ;) If you make it work in containerlab I could take a look into what would be needed to build a Vagrant box.

Now for the really important bit: what should we use as the device name? `vjunos-switch` is a bit long. Just `vjunos`?

> There might be some customization needed for the way vlans are handled, most probably, but all the other modules should work as is.

We can declare VLANs unsupported at the moment and leave them _for further study_ ;)

**Reply by ssasso** [reply]
I will try do to something in the next couple of weeks

**Reply by ssasso** [reply]
naming:

If I have to try with vjunos-switch, I'll try also for vjunos-router ("compact vMX - single VM").

See: https://containerlab.dev/manual/kinds/vr-vjunosswitch/ and https://containerlab.dev/manual/kinds/vr-vjunosrouter/

So I can propose:

- vjswitch
- vjrouter

otherwise, let's keep the long names

**Reply by ipspace** [reply]
"vjswitch" is awful. Let's go with vjunos-switch and see if the hyphen breaks anything.

**Comment by sjohnson1980** [comment]
Thanks Ivan for the quick responses. I would not mind trying it and helping
out. I was still debating around either reinstalling EVE, containerlab  or
possibly Netlab if it could work out on my lab BMS, as long as I would know
which direction I'd need to take to try it out and create the profile.

**Reply by ssasso** [reply]
> Thanks Ivan for the quick responses. I would not mind trying it and helping out. I was still debating around either reinstalling EVE, containerlab or possibly Netlab if it could work out on my lab BMS, as long as I would know which direction I'd need to take to try it out and create the profile.

Well, if I'm not wrong, EVE-ng distro is based on Ubuntu LTS, so **in theory** you could try installing containerlab && netlab over it ;)

**Comment by ipspace** [comment]
The integration tests are done, here's the scorecard:

https://tests.netlab.tools/_html/vjunos-switch-clab

As expected, the VLAN tests all failed, other failures are par-for-the-course considering the quality of our Junos implementation 🤷‍♂️

@sjohnson1980:

* The code is ready to go, run netlab from local copy of Git repository (https://netlab.tools/install/clone/) until I push out 1.9.0
* Send me a few simple working VLAN examples, we might be able to get it working with easy tweaks.

Finally, notice to anyone who might be reading this: we're accepting pull requests if you feel like fixing stuff ;)

**Reply by ssasso** [reply]
> other failures are par-for-the-course considering the quality of our Junos implementation

I know someone who could potentially help us on the JunOS config review/update ;) ( @nmodena puoi sentirti chiamato in causa :D )

**Reply by ipspace** [reply]
I don't think @nmodena responds to GitHub notifications 🥲 However, I plan to meet him in September 😎

**Reply by sjohnson1980** [reply]
Excellent, thanks, I hope to have time this week to have a look at it and also have other potential resources to help out

**Reply by ssasso** [reply]
> As expected, the VLAN tests all failed

@ipspace according to the documentation, the device should behave like the vPTX (with _interface family ethernet-switching_).

Could you please try the VLAN integration tests by using the same template (copy or symlink)? (there's no generic junos template on vlan module because each device class acts differently)

**Reply by ipspace** [reply]
Copied vPTX VLAN template into vJunos-switch template. L2 works, anything L3-related fails. Test results @ https://tests.netlab.tools/_html/vjunos-switch-clab-vlan

I checked the initial configuration template, and there's a suspicious "if device is not vPTX" condition there, so maybe that needs to be changed?
